### DISCUSSION TITLE
Option Greeks Missing from reqMktData() Response

### DISCUSSION THREAD

**Discussion by mauera** [lbilli/Jib.jl#32] [Q&A]
Hi, 
I am having trouble getting the greek values for an option contract (having already subscribed to the underlying). tickOptionComputation is not being called. Here is a snippet of my code used to generate the issue: 

```
trade_symbol = "SPY"
stock_contract = Jib.Contract(
        symbol=trade_symbol,
        secType="STK",
        exchange="SMART",
        currency="USD")
option_contract = Jib.Contract(
        symbol="SPY",
        secType = "OPT",
        exchange="SMART",
        currency="USD",
        conId=523596903
        )
option_contract.strike = 449.0
option_contract.lastTradeDateOrContractMonth = "20220520"
option_contract.right = "C"

Jib.reqMktData(ib_connect, 4, stock_contract, "", false, false)
Jib.reqMktData(ib_connect, 13, option_contract, "", false, false)
```

I  am assuming that the tickId input is  used to specify the type of tick data desired. Is this not the case?
Any suggestions would be appreciated. Thank you.

**Comment by lbilli** [comment]
This works:
```julia
opt = Jib.Contract(conId=523596903, exchange="SMART")

Jib.reqMktData(ib, 123, opt, "", false)
```
In order to get realtime market data you need a paid subscription. 
If you don't, you can still get delayed data if you run:
```julia
Jib.reqMarketDataType(ib, Jib.DELAYED)
```
before `reqMktData()`.

I think there's a misunderstanding regarding the unfortunately named argument `tickerId`: this is an arbitrary user defined unique integer used to identify TWS responses, it plays the same role as `reqId` in other requests. 
It has nothing to do with tick types.

In order to add tick types to a data stream, you take the value from the generic tick column and add it to the `genericTickList` argument. 
For example, if you want to add a stock historical and implied volatility to the usual pricing data:
```julia
Jib.reqMktData(ib, 1234, stock, "104,106", false)
```

**Reply by mauera** [reply]
Thank you for the help. I do have a real time data subscription. At one point the API was returning the Greeks, but subsequent calls would  not yield the same results. 
`"MODEL_OPTION", 0, 0.18242236317836005, 0.32285353456163435, 4.483781788251022, 0.0, 0.01650708199691522, 0.428788044918182, -0.1471393635893186, 438.06`

Now it just returns a subset of the expected tick types, and in both cases, the price is far from the expected value, 
` (tickerId = 2, minTick = 0.01, bboExchange = "c70003", snapshotPermissions = 0)
 (reqId = 2, marketDataType = Jib.REALTIME)
 (tickerId = 2, field = "BID", price = -1.0, size = 0.0, attrib = (canAutoExecute = true, pastLimit = false, preOpen = false)) (tickerId = 2, field = "ASK", price = -1.0, size = 0.0, attrib = (canAutoExecute = true, pastLimit = false, preOpen = false)) (tickerId = 2, field = "BID_SIZE", size = 0.0)
 (tickerId = 2, field = "ASK_SIZE", size = 0.0)
 (tickerId = 2, field = "LAST_SIZE", size = 0.0)
 (tickerId = 2, field = "CLOSE", price = 4.32, size = 0.0, attrib = (canAutoExecute = false, pastLimit = false, preOpen = false))`

I suspect the API will not return the expected values reliably outside of market hours, because I cannot replicate the earlier API responses.

**Reply by mauera** [reply]
When I used the suggested code, I get the following response, in aggregate, 
` (tickerId = 123, minTick = 0.01, bboExchange = "c70003", snapshotPermissions = 0)
 (reqId = 123, marketDataType = Jib.REALTIME)
 (tickerId = 123, field = "BID", price = -1.0, size = 0.0, attrib = (canAutoExecute = true, pastLimit = false, preOpen = false))
 (tickerId = 123, field = "ASK", price = -1.0, size = 0.0, attrib = (canAutoExecute = true, pastLimit = false, preOpen = false))
 (tickerId = 123, field = "BID_SIZE", size = 0.0)
 (tickerId = 123, field = "ASK_SIZE", size = 0.0)
 (tickerId = 123, field = "LAST_SIZE", size = 0.0)
 (tickerId = 123, field = "CLOSE", price = 4.32, size = 0.0, attrib = (canAutoExecute = false, pastLimit = false, preOpen = false))
 (tickerId = 123, tickType = "HALTED", value = 0.0)
 (tickerId = 123, tickType = "HALTED", value = 0.0)`

It does not appear to be invoking the tickOptionComputation callback.

**Reply by mauera** [reply]
![image](https://user-images.githubusercontent.com/682096/164948543-bbc0190b-48d9-4576-ac27-bc38734f6352.png)
There seems to be a requirement to subscribe to the underlying as well, according to the docs. I had done this before when I got the expected results back. This time it is not working.

**Reply by lbilli** [reply]
I think you should check back during market hours.

**Reply by mauera** [reply]
My thoughts as well, I might have gotten lucky  during the weekend on some price update, 
![image](https://user-images.githubusercontent.com/682096/164948639-26298c4e-093a-42a0-863e-4132993612d9.png)
This implies the callback is only triggered on a change in the underlying.

**Reply by mauera** [reply]
Thanks, works great, conId made the difference.
